embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/storage/PayloadStorageServiceImpl.kt

```diff
@@ -61,22 +84,42 @@ internal class PayloadStorageServiceImpl(
         }
     }
 
-    private fun pruneIfNeeded() {
-        // TODO: future: avoid calling listFiles() every time by retaining an in-memory list
-        val files = payloadDir.listFiles() ?: return
-        val metadataList = files.mapNotNull {
-            StoredTelemetryMetadata.fromFilename(it.name).getOrNull()
-        }.sortedWith(storedTelemetryComparator)
-
-        if (metadataList.size >= storageLimit) {
-            val excess = metadataList.size - storageLimit + 1
-            metadataList.takeLast(excess).forEach { it.asFile().delete() }
+    private fun pruneStorage(metadata: StoredTelemetryMetadata): Boolean {
+        val count = storedFiles.size
+        if (count < storageLimit) {
+            return false
+        }
+        val input = storedFiles.plus(metadata)
+        val removalCount = input.size - storageLimit
+        if (removalCount < 0) {
+            return false
         }
+        val removals = input
+            .sortedWith(storedTelemetryComparator)
+            .takeLast(removalCount)
+        removals.forEach(::delete)
+        internalErrorService.handleInternalError(RuntimeException("Pruned payload storage"))
```

Comment:
The semantics of this feel weird since it's not an error 😅 

Maybe we need to rename this later so we can relieve (my) cognitive dissonance

Reply:
Agree - it feels strange